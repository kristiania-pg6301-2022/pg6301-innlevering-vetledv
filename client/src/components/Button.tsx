interface TButton {
  clickHandler: (event: React.MouseEvent<HTMLButtonElement>) => void
}

export const Button: React.FC<TButton> = ({ children, clickHandler }) => {
  return (
    <div className='py-2'>
      <button
        onClick={clickHandler}
        className='px-6 py-3 font-light tracking-wide text-primary text-2xl capitalize transition-colors duration-200 transform bg-secondary rounded-md hover:bg-search-hover focus:outline-none focus:ring focus:ring-blue-300 focus:ring-opacity-80'>
        {children}
      </button>
    </div>
  )
}
