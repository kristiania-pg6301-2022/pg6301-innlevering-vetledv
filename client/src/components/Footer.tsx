import { ReactIcon } from './Icons'

export const Footer = () => {
  return (
    <footer className='h-24 bg-secondary w-full flex flex-row justify-evenly content-center'>
      <div className='flex py-7'>
        <p className='p-1 text-2xl'>Created using React.js</p>
        <ReactIcon />
      </div>
      <div className='py-7'>
        <a
          href='https://github.com/kristiania-pg6301-2022/pg6301-innlevering-vetledv'
          className='flex'>
          <p className='p-1 text-2xl'>GitHub</p>
          <img
            className='p-1 h-12 w-12'
            src='data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyNCIgaGVpZ2h0PSIyNCIgdmlld0JveD0iMCAwIDI0IDI0Ij48cGF0aCBkPSJNMTIgMGMtNi42MjYgMC0xMiA1LjM3My0xMiAxMiAwIDUuMzAyIDMuNDM4IDkuOCA4LjIwNyAxMS4zODcuNTk5LjExMS43OTMtLjI2MS43OTMtLjU3N3YtMi4yMzRjLTMuMzM4LjcyNi00LjAzMy0xLjQxNi00LjAzMy0xLjQxNi0uNTQ2LTEuMzg3LTEuMzMzLTEuNzU2LTEuMzMzLTEuNzU2LTEuMDg5LS43NDUuMDgzLS43MjkuMDgzLS43MjkgMS4yMDUuMDg0IDEuODM5IDEuMjM3IDEuODM5IDEuMjM3IDEuMDcgMS44MzQgMi44MDcgMS4zMDQgMy40OTIuOTk3LjEwNy0uNzc1LjQxOC0xLjMwNS43NjItMS42MDQtMi42NjUtLjMwNS01LjQ2Ny0xLjMzNC01LjQ2Ny01LjkzMSAwLTEuMzExLjQ2OS0yLjM4MSAxLjIzNi0zLjIyMS0uMTI0LS4zMDMtLjUzNS0xLjUyNC4xMTctMy4xNzYgMCAwIDEuMDA4LS4zMjIgMy4zMDEgMS4yMy45NTctLjI2NiAxLjk4My0uMzk5IDMuMDAzLS40MDQgMS4wMi4wMDUgMi4wNDcuMTM4IDMuMDA2LjQwNCAyLjI5MS0xLjU1MiAzLjI5Ny0xLjIzIDMuMjk3LTEuMjMuNjUzIDEuNjUzLjI0MiAyLjg3NC4xMTggMy4xNzYuNzcuODQgMS4yMzUgMS45MTEgMS4yMzUgMy4yMjEgMCA0LjYwOS0yLjgwNyA1LjYyNC01LjQ3OSA1LjkyMS40My4zNzIuODIzIDEuMTAyLjgyMyAyLjIyMnYzLjI5M2MwIC4zMTkuMTkyLjY5NC44MDEuNTc2IDQuNzY1LTEuNTg5IDguMTk5LTYuMDg2IDguMTk5LTExLjM4NiAwLTYuNjI3LTUuMzczLTEyLTEyLTEyeiIvPjwvc3ZnPg=='
          />
        </a>
      </div>
      <div className='py-7'>
        <a href='' className='flex'>
          <p className='p-1 text-2xl'>Heroku</p>
          <img
            className='h-12 w-12'
            src='https://www.vectorlogo.zone/logos/heroku/heroku-icon.svg'
          />
        </a>
      </div>
    </footer>
  )
}
