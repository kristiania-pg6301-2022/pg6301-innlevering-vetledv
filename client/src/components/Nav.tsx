import { Link } from 'react-router-dom'
import { useDarkMode } from '../hooks/useDarkMode'
import { ThemeBtn } from './ThemeBtn'

export const Nav = () => {
  const [colorTheme, setTheme] = useDarkMode()

  return (
    <div className='sticky top-0 w-full bg-primary p-5 flex justify-between shadow-md'>
      <Link to={'/'}>Home</Link>
      <ThemeBtn colorTheme={colorTheme} setTheme={setTheme} />
    </div>
  )
}
