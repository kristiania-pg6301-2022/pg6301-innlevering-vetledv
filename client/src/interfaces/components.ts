export interface TMutate {
  id: number
  answer: string
}

export type AnswerCorrect = {answerCorrect:boolean}