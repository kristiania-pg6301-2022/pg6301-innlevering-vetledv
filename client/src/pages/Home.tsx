import { Link, useNavigate } from 'react-router-dom'
import { Button } from '../components/Button'

export const Home = () => {
  const navigate = useNavigate()
  return (
    <>
      <Button clickHandler={() => navigate('/question')}>
        Get a random question
      </Button>
    </>
  )
}
