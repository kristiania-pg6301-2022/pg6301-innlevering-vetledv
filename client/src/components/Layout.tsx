import { Nav } from './Nav'
import { Footer } from './Footer'
export const Layout: React.FC = ({ children }) => {
  return (
    <>
        <div className='flex justify-between flex-col min-h-screen'>
      <div>
        <Nav />
        <main className='p-6'>{children}</main>
      </div>
      <Footer />
    </div>
      
    </>
  )
}
