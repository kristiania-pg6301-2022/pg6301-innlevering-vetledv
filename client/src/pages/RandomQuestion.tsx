import { useState } from 'react'
import { useMutation, useQuery, useQueryClient } from 'react-query'
import { Link, useNavigate } from 'react-router-dom'
import { AnswerCorrect, TMutate } from '../interfaces/components'
import { TQuestions } from '../interfaces/fetch'
import { useRandomQuestion } from './../hooks/useQuestions'
import { Button } from '../components/Button'

const postQuestions = async (url: string, json: Object) => {
  const res = await fetch(url, {
    method: 'post',
    headers: {
      'content-type': 'application/json',
    },
    body: JSON.stringify(json),
  })
  return res.json()
}

export const RandomQuestion = () => {
  const navigate = useNavigate()
  const queryClient = useQueryClient()
  const [answer, setAnswer] = useState<string>('')
  const query = useRandomQuestion('api/questions')
  const checkAnswer = useMutation<AnswerCorrect, Error, TMutate, () => void>(
    ({ id, answer }) => postQuestions('/api/questions', { id, answer })
  )

  const handleSubmit = async () => {
    if (answer !== '' && query.data) {
      const data = checkAnswer.mutate(
        {
          id: query.data.id,
          answer: answer,
        },
        {
          onSuccess: (answerRes) => {
            queryClient.setQueryData(['answer', query.data.id], answerRes)
          },
        }
      )
    }
  }

  return (
    <>
      <Button clickHandler={() => query.refetch()}>Get a new question</Button>
      <div>
        {query.isLoading && <div>Loading...</div>}
        {query.isError && <div>Error: {query.error.message}</div>}
        {query.data && (
          <div>
            <div>{query.data.question}</div>
            {Object.keys(query.data.answers)
              .filter((a) => query.data.answers[a])
              .map((a) => (
                <div key={a}>
                  <label>
                    <input
                      type='radio'
                      value={a}
                      name='answer'
                      onChange={(event) => {
                        setAnswer(event.target.value)
                      }}
                    />
                    {query.data.answers[a]}
                  </label>
                </div>
              ))}
            <div className='flex gap-2'>
              <Button clickHandler={() => handleSubmit()}>Submit</Button>
              <Button clickHandler={() => navigate('/')}>Back</Button>
            </div>
            {checkAnswer.isSuccess && (
              <div>
                {checkAnswer.data.answerCorrect
                  ? 'Correct!'
                  : 'Wrong, try again'}
              </div>
            )}
          </div>
        )}
      </div>
    </>
  )
}
