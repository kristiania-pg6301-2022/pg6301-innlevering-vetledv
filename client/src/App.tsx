import { QueryClient, QueryClientProvider } from 'react-query'
import { ReactQueryDevtools } from 'react-query/devtools/development'
import { BrowserRouter, Route, Routes } from 'react-router-dom'
import { Layout } from './components/Layout'
import { Home } from './pages/Home'
import { RandomQuestion } from './pages/RandomQuestion'

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      refetchOnMount: true,
      keepPreviousData: true,
    },
  },
})

const App = () => {
  return (
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <Layout>
          <Routes>
            <Route path='/' element={<Home />} />
            <Route path='/question' element={<RandomQuestion />} />
            <Route path='*' element={<div>404</div>} />
          </Routes>
        </Layout>
      </BrowserRouter>
      <ReactQueryDevtools initialIsOpen />
    </QueryClientProvider>
  )
}

export default App
