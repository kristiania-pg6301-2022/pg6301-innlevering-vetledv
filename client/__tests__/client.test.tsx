import { renderHook } from '@testing-library/react-hooks'
import { render, screen, waitFor } from '@testing-library/react'
import pretty from 'pretty'
import { render as reactdomrender, unmountComponentAtNode } from 'react-dom'
import { QueryClient, QueryClientProvider, setLogger } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import App from '../src/App'
import { Layout } from '../src/components/Layout'
import { Home } from '../src/pages/Home'
import { RandomQuestion } from '../src/pages/RandomQuestion'
import { rest, server } from '../src/mocks/server'
require('whatwg-fetch')
import '@testing-library/jest-dom'
import { id } from '../jest.setup'

const createTestQueryClient = () =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

export const createWrapper = () => {
  const testQueryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })
  return ({ children }: { children: React.ReactElement }) => (
    <QueryClientProvider client={testQueryClient}>
      {children}
    </QueryClientProvider>
  )
}
export function renderWithClient(ui: React.ReactElement) {
  const testQueryClient = createTestQueryClient()
  const { rerender, ...result } = render(
    <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
  )
  return {
    ...result,
    rerender: (rerenderUi: React.ReactElement) =>
      rerender(
        <QueryClientProvider client={testQueryClient}>
          {rerenderUi}
        </QueryClientProvider>
      ),
  }
}

describe('quiz', () => {
  //nock bad, msw good :)
  let container: HTMLDivElement

  beforeEach(async () => {
    container = await document.createElement('div')
  })
  afterEach(async () => {
    unmountComponentAtNode(await container)
  })

  test('render Layout', () => {
    reactdomrender(
      <MemoryRouter>
        <Layout></Layout>
      </MemoryRouter>,
      container
    )
    expect(pretty(container.innerHTML)).toMatchSnapshot()
  })
  test('render App', () => {
    reactdomrender(<App />, container)
    expect(pretty(container.innerHTML)).toMatchSnapshot()
  })

  test('render page Home', () => {
    reactdomrender(
      <MemoryRouter>
        <Home />
      </MemoryRouter>,
      container
    )
    expect(pretty(container.innerHTML)).toMatchSnapshot()
  })

  test('render page RandomQuestion', () => {
    const queryClient = new QueryClient()
    reactdomrender(
      <QueryClientProvider client={queryClient}>
        <MemoryRouter>
          <RandomQuestion />
        </MemoryRouter>
      </QueryClientProvider>,
      container
    )
    expect(pretty(container.innerHTML)).toMatchSnapshot()
  })
  test('loading on page RandomQuestion', async () => {
    renderWithClient(
      <MemoryRouter>
        <RandomQuestion />
      </MemoryRouter>
    )
    expect(await screen.findByText(/Loading.../i)).toBeInTheDocument()
  })
  test('error on page RandomQuestion', async () => {
    server.use(
      rest.get('*', (req, res, ctx) => {
        return res(ctx.status(500))
      })
    )
    renderWithClient(
      <MemoryRouter>
        <RandomQuestion />
      </MemoryRouter>
    )

    expect(await screen.findByText(/Error:/i)).toBeInTheDocument()
  })
  test('RandomQuestion page displays data', async () => {
    renderWithClient(
      <MemoryRouter>
        <RandomQuestion />
      </MemoryRouter>
    )
    await waitFor(() => {
      expect(
        screen.getByText(/Good day sir what's your name sir/i)
      ).toBeInTheDocument()
    })
  })
})
