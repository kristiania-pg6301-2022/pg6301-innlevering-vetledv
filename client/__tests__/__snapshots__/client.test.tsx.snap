// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`quiz should render Home component 1`] = `"<button class=\\"px-4 py-2 font-medium tracking-wide text-white capitalize transition-colors duration-200 transform bg-blue-600 rounded-md hover:bg-blue-500 focus:outline-none focus:ring focus:ring-blue-300 focus:ring-opacity-80\\">Get a random question</button>"`;
